src/raw_client.rs

```diff
@@ -668,32 +667,32 @@ impl<T: Read + Write> ElectrumApi for RawClient<T> {
 
         self.increment_calls();
 
-        while !missing_responses.is_empty() {
-            let req_id = *missing_responses.iter().next().unwrap();
-            let resp = match self.recv(&receiver, req_id) {
-                Ok(resp) => resp,
+        let mut answers = BTreeMap::new();
+        for req_id in missing_responses {
+            match self.recv(&receiver, req_id) {
                 Err(e) => {
                     // In case of error our sender could still be left in the map, depending on where
                     // the error happened. Just in case, try to remove it here
                     warn!("got error for req_id {}: {:?}", req_id, e);
                     warn!("removing all waiting req of this batch");
-                    let mut guard = self.waiting_map.lock().unwrap();
-                    for req_id in missing_responses.iter() {
-                        guard.remove(&req_id);
-                    }
+                    let mut guard = self
+                        .waiting_map
+                        .lock()
+                        .map_err(|_| Error::CouldntLockReader)?;
+                    guard.remove(&req_id);
                     return Err(e);
                 }
-            };
-            let resp_id = resp["id"].as_u64().unwrap() as usize;
-
-            missing_responses.remove(&resp_id);
-            answer.push(resp);
+                Ok(resp) => {
+                    let mut res = resp
+                        .get("result")
+                        .cloned()
```

Comment:
I would keep this part as it was before, just `take()`-ing the result and adding it to `answer`.

There's no need to return an `InvalidResponse` here because if there's something wrong with it we'll just see that later when trying to parse the response into a specific struct.

I guess if you really wanted to return that error instead of a generic `serde` error you could add this to `impl_batch_call`:

 https://github.com/bitcoindevkit/rust-electrum-client/blob/af10aeac99bcd2b68c1b84b264d0971c76873301/src/raw_client.rs#L49

You could map the error here to `InvalidResponse` before returning it